### PR TITLE
Various fixes

### LINKED ISSUE #9: Matching groups
It would be useful to be able to get organization name, repository name, and issue number separately from the match result. This just requires adding three capturing groups to the regex.

### LINKED ISSUE #10: Doesn't accept single-letter org or repo names
Repo names of length one are valid -- easy to see by creating one in your org.
Organisation names of length one are also valid, for example https://github.com/a/

Both aren't accepted it seems:
```js
> /(?:(?<![/\w-.])\w[\w-.]+?\/\w[\w-.]+?|\B)#[1-9]\d*?\b/g.exec('foo/bar#123')
[ 'foo/bar#123', index: 0, input: 'foo/bar#123', groups: undefined ]
> /(?:(?<![/\w-.])\w[\w-.]+?\/\w[\w-.]+?|\B)#[1-9]\d*?\b/g.exec('foo/a#123')
null
> /(?:(?<![/\w-.])\w[\w-.]+?\/\w[\w-.]+?|\B)#[1-9]\d*?\b/g.exec('a/foo#123')
null
```

### LINKED ISSUE #11: Shouldn't match overly long names
GitHub has a 100 character limit for repository names:
Screen Shot 2022-03-20 at 22 10 40

And a 39 character limit for organization names:
Screen Shot 2022-03-20 at 22 12 31

I can't find info on the max number of issues per repository, but it probably makes sense to cap the issue number as well. It's hard to imagine a repository having more than -- say -- a billion issues 😄 

The regex would ideally reflect this. Right now any length will match:

```js
> !!/(?:(?<![/\w-.])\w[\w-.]+?\/\w[\w-.]+?|\B)#[1-9]\d*?\b/g.exec(`${"o".repeat(1000)}/${"r".repeat(1000)}#${"1".repeat(1000)}`)
true
```

Aside from avoiding to match gibberish, doing this would also help slightly with performance. Here's an unscientific benchmark.

```js
> function benchmark(f) { const a = Date.now(); f(); console.log(Date.now() - a); }
undefined
> benchmark(() => /(?:(?<![/\w-.])\w[\w-.]+?\/\w[\w-.]+?|\B)#[1-9]\d*?\b/g.exec("x".repeat(100000000)))
802
undefined
> benchmark(() => /(?:(?<![/\w-.])\w[\w-.]{1,100}?\/\w[\w-.]{0,100}?|\B)#[1-9]\d{0,100}?\b/g.exec("x".repeat(100000000)))
553
undefined
```

(Note: the regex above is just an example to test performance roughly, it's not meant to be correct.)

### LINKED ISSUE #12: Missing Prettier configuration?
It seems that code style is enforced by `xo` and it uses `prettier`, and sets a version (2.6.0) for it in `yarn.lock`. But running `yarn && yarn prettier --write '*.js' && yarn test` in this repository produces [errors](https://gist.github.com/jscheid/2327647745ee9901fd56a06fc54ab73c). Is it perhaps expecting a non-default Prettier configuration? If so, it would be good to add that to the repository because otherwise it's extremely annoying to have to manually keep the linter happy.

### LINKED ISSUE #14: Edge cases
Unlike a repository name, an org name cannot contain underscores...
Screen Shot 2022-03-21 at 09 30 31

... or dots:
Screen Shot 2022-03-21 at 09 30 44

And, like the error message ☝️ says, a hyphen can only come in the middle of the name, not at the beginning or end.

(I've checked and the same limitations, including the 39-character limit, apply to GitHub _user names_ as well, not just organizations.)

For repository names, it's ok to start with a hyphen -- in fact, a repository can be named `-`: https://github.com/jscheid/-

A repository name can start with a dot: https://github.com/jscheid/.foo

... but it can't be just a single dot (or two dots) for obvious reasons:

Screen Shot 2022-03-21 at 09 35 03

It's ok for `org/repo#num` to follow a slash: foo/bar/sindresorhus/issue-regex#13

It's ok for `org/repo#num` to follow a dot: foobar.sindresorhus/issue-regex#13

It's ok for `org/repo#num` to follow a dash: -sindresorhus/issue-regex#13
... but not `foo-` because that would be a different org: foo-sindresorhus/issue-regex#13

It's not ok for `org/repo#num` to follow an underscore, even though underscores aren't allowed in org names: foo_sindresorhus/issue-regex#13

### PR DESCRIPTION
**Pull request by jscheid**
- Match single-character organisation or repository name
- Don't match organisation or repository name that are too long
- Don't match issue numbers greater than 10^100-1
- Put organisation name, repository name and issue number into capturing groups 1, 2 and 3.
- Add missing Prettier configuration

Fixes #9, Fixes #10, Fixes #11, Fixes #12, Fixes #14

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
 // https://regex101.com/r/SQrOlx/14
 export default function issueRegex() {
-	return /(?:(?<![/\w-.])\w[\w-.]+?\/\w[\w-.]+?|\B)#[1-9]\d*?\b/g;
+	return /(?<![\w./-])(?:(?<org>\w[\w.-]{0,38})\/(?<repo>\w[\w.-]{0,99}))?#(?<num>[1-9]\d{0,99})\b/g;
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	return /(?<![\w./-])(?:(?<organization>\w[\w.-]{0,38})\/(?<repo>\w[\w.-]{0,99}))?#(?<issueNumber>[1-9]\d{0,99})\b/g;
```

File: test.js [hunk 2]
```diff
@@ -20,6 +25,9 @@ const nonMatches = [
 	'#123hashtag',
 	'non/-repo#123',
 	'this/is/not/repo#123',
+	'thisorganisationnameistoolongxxxxxxxxxxx/foo#123',
+	'foo/thisrepositorynameistoolongxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx#123',
+	'foo/thisissuenumberistoolong#11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111',
```
Comment (sindresorhus) [inline test.js]: Can you add a code comment about why these are non-match. Like that the max length is this much.

**Comment by yakov116** [comment]
> Fixes #9, #10, #11, #12

You need to do Fixes #9, Fixes #10 etc for GitHub to close the issue

**Comment by jscheid** [comment]
Perhaps we shouldn't merge this until we've decided what to do with #15. If that gets implemented, the capturing groups might change and necessitate a release with breaking changes. Probably better to sort this all out in one go.

File: index.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
 // https://regex101.com/r/SQrOlx/14
 export default function issueRegex() {
-	return /(?<![\w./-])(?:(?<organization>\w[\w.-]{0,38})\/(?<repository>\w[\w.-]{0,99}))?#(?<issueNumber>[1-9]\d{0,99})\b/g;
+	return /(?<!\w)(?:(?<organization>[a-z\d](?:[a-z\d-]{0,37}[a-z\d])?)\/(?<repository>[\w.-]{1,100}))?(?<!(?:\/\.{1,2}))#(?<issueNumber>[1-9]\d{0,99})\b/gi;
```
Comment (sindresorhus) [inline index.js]: Let's limit the issue number length to 10. There will never be 9 billion issues in a repo ever.

**Comment by jscheid** [comment]
On second thought, why don't we merge this and then address #15 in a separate PR, and wait with a release until after that?

**Comment by sindresorhus** [comment]
> then address https://github.com/sindresorhus/issue-regex/issues/15 in a separate PR, and wait with a release until after that?

👍

**Comment by sindresorhus** [comment]
> On second thought, why don't we merge this and then address https://github.com/sindresorhus/issue-regex/issues/15 in a separate PR, and wait with a release until after that?

@jscheid Still interested?

**Comment by jscheid** [comment]
Vaguely, but perhaps you should go ahead with a release if that's what you're asking?

**Pull request merged**